Source/Scene/ParticleSystem.js

```diff
@@ -0,0 +1,295 @@
+        particle.startScale = this.startScale;
+        particle.endScale = this.endScale;
+        particle.image = this.image;
+        particle.life = random(this.minLife, this.maxLife);
+        particle.mass = random(this.minMass, this.maxMass);
+
+        particle.size = new Cartesian2(random(this.minWidth, this.maxWidth), random(this.minHeight, this.maxHeight));
+
+        var speed = random(this.minSpeed, this.maxSpeed);
+        Cartesian3.multiplyByScalar(particle.velocity, speed, particle.velocity);
+
+        this.particles.push(particle);
+    };
+
+    ParticleSystem.prototype.calcNumberToEmit = function(dt) {
+        // This emitter is finished if it exceeds it's lifetime.
+        if (this.isComplete) {
+            return 0;
+        }
+
+        // Compute the number of particles to emit based on the rate.
+        var v = dt * this.rate;
+        var numToEmit = Math.floor(v);
+        this.carryOver += (v-numToEmit);
+        if (this.carryOver>1.0)
+        {
+            numToEmit++;
+            this.carryOver -= 1.0;
+        }
+
+
+        var i = 0;
+
+        // Apply any bursts
+        if (this.bursts) {
+            for (i = 0; i < this.bursts.length; i++) {
+                var burst = this.bursts[i];
+                if ((!defined(burst, "complete") || !burst.complete) && this.currentTime > burst.time) {
+                    var count = burst.min + random(0.0, 1.0) * burst.max;
+                    numToEmit += count;
+                    burst.complete = true;
+                }
+            }
+        }
+
+        return numToEmit;
+    };
+
+    ParticleSystem.prototype.update = function(frameState) {
+        if (!defined(this._billboardCollection)) {
+            this._billboardCollection = new BillboardCollection();
+        }
+
+        // Compute the frame time
+        var dt = 0.0;
+        if (this._previousTime) {
+            dt = JulianDate.secondsDifference(frameState.time, this._previousTime);
+        }
+
+        if (dt < 0.0) {
+            dt = 0.0;
+        }
+
+        var i = 0;
+
+
+        var particles = this.particles;
+        var emitter = this.emitter;
+
+        var particle = null;
+
+        // update particles and remove dead particles
+        var length = particles.length;
+        for (i = 0; i < length; ++i) {
+            particle = particles[i];
+            if (!particle.update(this.forces, dt)) {
+                removeBillboard(this, particle);
+                particles[i] = particles[length - 1];
+                --i;
+                --length;
+            } else {
+                updateBillboard(this, particle);
+            }
+        }
+        particles.length = length;
+
+
+        var numToEmit = this.calcNumberToEmit(dt);
+
+        if (numToEmit > 0 && emitter) {
+
+            // Compute the final model matrix by combining the particle systems model matrix and the emitter matrix.
+            var combinedMatrix = new Matrix4();
```

Comment:
This and the tmp variable are converted to scratch vars in 8a8626cd61e26e2f1f19bd60c1bb6866b613d1a4